driver/src/price_estimation/dexag/api.rs

```diff
@@ -0,0 +1,183 @@
+#![allow(dead_code)]
+
+use anyhow::{anyhow, Context, Result};
+use ethcontract::Address;
+use isahc::prelude::*;
+use serde::{de::Error as _, Deserialize, Deserializer};
+use serde_with::rust::display_fromstr;
+use std::time::Duration;
+use url::Url;
+
+#[cfg_attr(test, mockall::automock)]
+/// Parts of the dex.ag api.
+pub trait DexagApi {
+    /// https://docs.dex.ag/api/tokens
+    fn token_list_full(&self) -> Result<Vec<Token>>;
+    /// https://docs.dex.ag/api/price
+    fn price(
+        &self,
+        from: &str,
+        to: &str,
+        amount: EitherAmount,
+        dex: Dex,
+        discluded: &str,
+    ) -> Result<Price>;
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct Token {
+    pub name: String,
+    pub symbol: String,
+    pub address: Option<Address_>,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct Price {
+    #[serde(with = "display_fromstr")]
+    pub price: f64,
+}
+
+/// Wrapper type of Address that implements deserialize from a hex string.
+#[derive(Clone, Debug)]
+pub struct Address_(pub Address);
+
+impl<'de> Deserialize<'de> for Address_ {
```

Comment:
The existing Deserialise or FromStr are not working?

Reply:
The existing deserialize expects the address hex string to not have the `0x` prefix. I haven't checked FromStr but I am going to.